### PR TITLE
[WIP] Use connectToExtensions HOC for reactive extension consumption

### PR DESCRIPTION
**Pull request by vojtechszocs**
Fixes: CNV-1772, CNV-1773

:construction_worker_man: **Work In Progress:** ~~needs rebase~~, ~~ESLint fixups~~, rework default perspective handling :construction_worker_woman:

This PR introduces `connectToExtensions` [higher-order component](https://reactjs.org/docs/higher-order-components.html) (HOC) used to connect existing components to Console extensions. This is a direct analogy to `connect` from [react-redux](https://github.com/reduxjs/react-redux), but using `Extension[]` as the source of truth.

Let's take `public/components/dashboards-page/dashboards.tsx` as an example.

<details><summary>code before</summary>
<p>

```tsx
import { connect } from 'react-redux';

const mapStateToProps = ({ k8s }) => ({
  kindsInFlight: k8s.getIn(['RESOURCES', 'inFlight']),
  k8sModels: k8s.getIn(['RESOURCES', 'models']),
});

const DashboardsPage_: React.FC<DashboardsPageProps> = ({
  match,
  kindsInFlight,
  k8sModels,
}) => {
  // call plugins.registry.getDashboardsCards()
  // call plugins.registry.getDashboardsTabs()
};

export const DashboardsPage = connect(mapStateToProps)(DashboardsPage_);

type DashboardsPageProps = RouteComponentProps & {
  kindsInFlight: boolean;
  k8sModels: ImmutableMap<string, any>;
};
```
</p>
</details>

<details><summary>code after</summary>
<p>

```tsx
import { connect } from 'react-redux';

import {
  connectToExtensions,
  Extension,
  DashboardsCard,
  DashboardsTab,
  isDashboardsCard,
  isDashboardsTab,
} from '@console/plugin-sdk';

const mapStateToProps = ({ k8s }) => ({
  kindsInFlight: k8s.getIn(['RESOURCES', 'inFlight']),
  k8sModels: k8s.getIn(['RESOURCES', 'models']),
});

const mapExtensionsToProps = (extensions: Extension[]) => ({
  pluginTabs: extensions.filter(isDashboardsTab),
  pluginCards: extensions.filter(isDashboardsCard),
});

const DashboardsPage_: React.FC<DashboardsPageProps> = ({
  match,
  kindsInFlight,
  k8sModels,
  pluginTabs,
  pluginCards,
}) => {
  // no call to plugins.registry
};

export const DashboardsPage = connect(mapStateToProps)(
  connectToExtensions(mapExtensionsToProps)(DashboardsPage_)
);

type DashboardsPageProps = RouteComponentProps & {
  kindsInFlight: boolean;
  k8sModels: ImmutableMap<string, any>;
  pluginCards: DashboardsCard[];
  pluginTabs: DashboardsTab[];
};
```
</p>
</details>

Components no longer reference the `plugins.registry` object directly. Instead, they specify how extensions **that are currently in use** are mapped to their props via `mapExtensionsToProps` function.

## Extensions in use

An extension is in use (takes effect) when:

- it is an always-on extension (i.e. not gated by any feature flags)

- otherwise, its `flags` constraints must be satisfied:

  - all `required` flags are resolved to `true`
  - all `disallowed` flags are resolved to `false`

```tsx
export type Extension<P = any> = {
  type: string;
  properties: P;
  flags?: Partial<{
    required: string[];
    disallowed: string[];
  }>;
};

export type AlwaysOnExtension<P = any> = Omit<Extension<P>, 'flags'>;
```

Extensions declared as always-on:
- `ModelFeatureFlag` - adding new feature flag based on CRD
- `ModelDefinition` - adding new (static) k8s model definitions

## Gating all plugin's extensions

`ModelFeatureFlag` extension type has a new `gateExtensions` property, which defaults to `true`.

When enabled, the corresponding flag is automatically added to `flags.required` array for **all the plugin's extensions** (excluding always-on extensions).

The default value of `true` reflects the general assumption that extensions contributed by plugin X should be gated by flag(s) introduced by plugin X.

## Gating specific extensions

Each extension can specify feature flags which are required and/or disallowed in order to put that particular extension into effect.

For example:

```ts
{
  type: 'NavItem/Href',
  properties: {
    perspective: 'dev',
    componentProps: { /* stuff */ },
  },
  flags: {
    required: [FLAGS.OPENSHIFT],
  },
},
```

At runtime, above DevConsole extension's required flags will become:

```ts
['OPENSHIFT', 'SHOW_PIPELINE']
```

due to a `FeatureFlag/Model` extension adding the `SHOW_PIPELINE` flag.

---

/cc @spadgett @alecmerdler @christianvogt @jelkosz @rawagner

### CODE REVIEW

**Comment by rawagner** [comment]
great work @vojtechszocs !

I wonder if it would be beneficial to store extensions to redux store and then the example you posted would look like

```javascript
import { connect } from 'react-redux';

import {
  connectToExtensions,
  Extension,
  DashboardsCard,
  DashboardsTab,
  isDashboardsCard,
  isDashboardsTab,
} from '@console/plugin-sdk';

const mapStateToProps = ({ k8s, extensions }) => ({
  kindsInFlight: k8s.getIn(['RESOURCES', 'inFlight']),
  k8sModels: k8s.getIn(['RESOURCES', 'models']),
  pluginTabs: extensions.filter(isDashboardsTab),
  pluginCards: extensions.filter(isDashboardsCard),
});

const DashboardsPage_: React.FC<DashboardsPageProps> = ({
  match,
  kindsInFlight,
  k8sModels,
  pluginTabs,
  pluginCards,
}) => {
  // no call to plugins.registry
};

export const DashboardsPage = connect(mapStateToProps)(DashboardsPage_);

type DashboardsPageProps = RouteComponentProps & {
  kindsInFlight: boolean;
  k8sModels: ImmutableMap<string, any>;
  pluginCards: DashboardsCard[];
  pluginTabs: DashboardsTab[];
};
```

is there any downside to this approach ?

**Comment by vojtechszocs** [comment]
Rebased on latest `master` and squashed some commits.

**Comment by vojtechszocs** [comment]
Fixed all ESLint issues.

**Comment by vojtechszocs** [comment]
Thanks @rawagner for your review! :smiley:

> I wonder if it would be beneficial to store extensions in redux store

`connectToExtensions` HOC creator is actually doing two things:
- compute extensions currently in use, i.e. `pluginStore.getExtensionsInUse(flags)`
- call `mapExtensionsToProps` and pass those additional props to wrapped `Component`

This is similar to existing HOC creators like `connectToFlags`.

Removing `connectToExtensions` in favor of using standard react-redux `connect` directly, we shift the responsibility of performing related computations to consumers:

```tsx
import {
  getExtensionsInUse,
  isDashboardsTab,
  isDashboardsCard,
} from '@console/plugin-sdk';

// in this example, "extensions" refers to *all* extensions stored in Redux
const mapStateToProps = ({ k8s, extensions }) => ({
  kindsInFlight: k8s.getIn(['RESOURCES', 'inFlight']),
  k8sModels: k8s.getIn(['RESOURCES', 'models']),
  pluginTabs: getExtensionsInUse(extensions.filter(isDashboardsTab)),
  pluginCards: getExtensionsInUse(extensions.filter(isDashboardsCard)),
});
```

but this could be optimized by storing `extensionsInUse` in Redux as _derived data_, updated whenever `extensions` (source of truth) is changed.

```tsx
const mapStateToProps = ({ k8s, extensionsInUse }) => ({
  kindsInFlight: k8s.getIn(['RESOURCES', 'inFlight']),
  k8sModels: k8s.getIn(['RESOURCES', 'models']),
  pluginTabs: extensionsInUse.filter(isDashboardsTab),
  pluginCards: extensionsInUse.filter(isDashboardsCard),
});
```

In terms of a reducer function (i.e. `public/reducers/extensions.ts`), the initialization of related sub-state (first argument of reducer) would depend on code like:

```ts
const activePlugins = (process.env.NODE_ENV !== 'test')
  ? require('@console/active-plugins').default as ActivePlugin[]
  : [];
```

breaking the constraint of reducer being a pure function. But we could work around this by passing pre-initialized `extensions` to `createStore` function, shifting the impurity to code outside the reducer:

```ts
// the second argument, "{}", is the initial Redux state
const store = createStore(reducers, {}, composeEnhancers(applyMiddleware(thunk)));
```

### TL;DR

I'd favor more specific HOC creators like `connectToFlags` and `connectToExtensions` over using the generic `connect`, since the latter forces us to
- either represent the computed data as derived data stored & managed in Redux
  - adds complexity to reducer functions
- or shift the responsibility of computing data to consumers
  - spreads complexity over the codebase

As for **how to store Console extensions**, I'm fine with having them managed via Redux state too.

I'm also curious what other people think about this.

**Comment by vojtechszocs** [comment]
@christianvogt @spadgett @jelkosz 

The only remaining issue before removing the WIP label is
- rework default perspective handling (part of 1fea92544ae8dc2c00fe42871abfbfc30ffdd929)

Reducer in `public/reducers/ui.ts` calls `plugins.registry.getPerspectives()` as part of the sub-state (`state: UIState`) initialization. This call is sync, imposing the assumption that extensions of type `Perspective` are available as early as during Console startup.

However, by design, only `ModelFeatureFlag` and `ModelDefinition` extensions are always-on. With the current code in `packages/dev-console/src/plugin.tsx`, the extension for "Developer" perspective is implicitly gated by the `SHOW_PIPELINE` flag.

_Note: the default "Administrator" perspective (`packages/console-app/src/plugin.tsx`) isn't gated by any flags, and therefore in use right upon Console startup._

So what's the best way to solve this problem?
- postpone Console startup until flags used to gate perspective extensions are resolved (UX impact)
- remove default perspective handling
- any other solution?

**Comment by vojtechszocs** [comment]
FYI @mareklibra @jtomasek @knowncitizen @honza @cloudbehl @joshuawilson @rohitkrai03

**Comment by knowncitizen** [comment]
Thanks for the HU on this @vojtechszocs

**Comment by vojtechszocs** [comment]
> My biggest concern with this change is that we might be introducing performance regressions and unnecessary renders when using the HOC with `React.memo`. I personally would prefer an approach that uses hooks. Not all components that consume extensions are functional components, however, so it might not be practical.

I can add a React hook as an alternative to `connectToExtensions` HOC for use with functional components & `React.memo`. I agree that we should avoid unnecessary performance overhead.

File: frontend/packages/console-plugin-sdk/src/connect.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * - otherwise, its `flags` constraints must be satisfied
+ *   - all required feature flags are resolved to `true`
+ *   - all disallowed feature flags are resolved to `false`
+ */
+export const connectToExtensions: ConnectToExtensions = (mapExtensionsToProps) => (Component) => {
+  const mapStateToProps = (state: RootState) => ({
+    allFlags: state[featureReducerName],
+  });
+
+  const ComponentWrapper = connect(mapStateToProps)((props) => {
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/connect.tsx]: :+1: will do.

File: frontend/packages/console-plugin-sdk/src/store.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  Extension,
+  AlwaysOnExtension,
+  ActivePlugin,
+  isFeatureFlag,
+  isModelDefinition,
+} from './typings';
+import { ExtensionRegistry } from './registry';
+
+const isAlwaysOnExtension = (e: Extension): e is AlwaysOnExtension => {
+  return isFeatureFlag(e) || isModelDefinition(e);
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/store.ts]: Agreed, will do.

The `isAlwaysOnExtension` implementation follows type declarations - both `FeatureFlag` and `ModelDefinition` extend `AlwaysOnExtension` (instead of `Extension`).

[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/store.ts]: I can also move `isAlwaysOnExtension` function (which is technically a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)) closer to extension type declarations, i.e. `packages/console-plugin-sdk/src/typings/index.ts` - let me know if I should.

File: frontend/packages/console-plugin-sdk/src/store.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+};
+
+const isGateableExtension = (e: Extension) => !isAlwaysOnExtension(e);
+
+const getGatingFlags = (p: ActivePlugin) => {
+  return [
+    ...p.extensions
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/store.ts]: You are right :smiley: I knew you'd ask about this.

The reason I wrote it that way is to allow easy expansion in the future, e.g.

```ts
return [
  ...p.extensions
    .filter(isFeatureFlag)
    .filter((e) => e.properties.gateExtensions === undefined || e.properties.gateExtensions)
    .map((e) => e.properties.flag),
  ...p.extensions
    .filter(isSomethingThatAffectsCrossExtensionFeatureGating)
    .filter((e) => e.properties.foo === 'whatever')
    .map((e) => e.properties.flag),
];
```

But maybe I'm over-thinking it. I'll switch to direct `filter` / `map` as suggested.

File: frontend/packages/console-plugin-sdk/src/typings/features.ts [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
 
 namespace ExtensionProperties {
   export interface ModelFeatureFlag {
-    /** If a CRD for this model exists, the feature will be enabled. */
+    /** If a CRD for this model exists, the feature flag will be enabled. */
     model: K8sKind;
     /** The name of the feature flag. */
     flag: string;
+    /** Whether to gate all of the plugin's extensions by this feature flag. Defaults to `true`. */
+    gateExtensions?: boolean;
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/typings/features.ts]: Agreed, will do.

Different plugins will likely use a different value, depending on the specific flag. For example, DevConsole generally doesn't want to gate its extensions by any flags, whereas KubeVirt generally wants to gate all of its extensions by the `KUBEVIRT` flag.

(*) its extensions = ones which are gateable (excluding always-on)

File: frontend/public/components/app-contents.tsx [hunk 5]
```diff
@@ -92,142 +93,154 @@ const LazyRoute = (props) => (
...
-          <LazyRoute loader={() => import('./error' /* webpackChunkName: "error" */).then(m => m.ErrorPage404)} />
-        </Switch>
+});
+
+const mapExtensionsToProps = (extensions: Extension[]) => ({
+  pluginRoutePages: extensions.filter(isRoutePage),
+});
+
+// use `withRouter` to force a re-render when routes change since we are using React.memo
+const AppContents = connect(mapStateToProps)(connectToExtensions(mapExtensionsToProps)(withRouter<AppContentsProps>(
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/public/components/app-contents.tsx]: Will check this and all similar usages.

File: frontend/public/components/dashboard/inventory-card/inventory-item.tsx [hunk 6]
```diff
@@ -62,32 +52,47 @@ export const InventoryItem: React.FC<InventoryItemProps> = React.memo(
...
-const StatusLink: React.FC<StatusLinkProps> = React.memo(({groupID, count, statusIDs, kind, namespace, filterType}) => {
-  const statusItems = encodeURIComponent(statusIDs.join(','));
-  const namespacePath = namespace ? `ns/${namespace}` : 'all-namespaces';
-  const to = filterType && statusItems.length > 0 ? `/k8s/${namespacePath}/${kind.plural}?rowFilter-${filterType}=${statusItems}` : `/k8s/${namespacePath}/${kind.plural}`;
-  const statusGroupIcons = getStatusGroupIcons();
-  const groupIcon = statusGroupIcons[groupID] || statusGroupIcons[InventoryStatusGroup.NOT_MAPPED];
-  return (
-    <div className="co-inventory-card__status">
-      <Link to={to} style={{textDecoration: 'none'}}>
+export const Status = React.memo(connectToExtensions(mapExtensionsToProps)(
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply frontend/public/components/dashboard/inventory-card/inventory-item.tsx]: Will check.

Anyway, we can use a  React hook instead of HOC here.

**Comment by christianvogt** [comment]
@vojtechszocs I dislike the behavior of `gateExtensions`. It's too magical. I would prefer to be explicit. You've introduced a `flags` prop to the extension, let the author be explicit and supply the flag even if it seems redundant.

**Comment by vojtechszocs** [comment]
Note to self (cc @christianvogt): remove `ModelFeatureFlag.gateExtensions` property, let people be explicit which flags are required per each extension.

**Comment by vojtechszocs** [comment]
#### Original action items
- [ ] rework default perspective handling as per https://github.com/openshift/console/pull/2269#issuecomment-518782910
- [ ] remove `ModelFeatureFlag.gateExtensions` property as per https://github.com/openshift/console/pull/2269#issuecomment-522699912
- [ ] add `useExtensions` hook to complement `connectToExtensions` HOC in functional components

#### Additional action items
- [ ] rebase on #2446 & remove any extension-specific gating properties like `required: string`
- [ ] address TODO items where the `react-hooks/exhaustive-deps` ESLint rule is disabled
- [ ] avoid using the `useEffectDeepCompare` hook (or similar custom comparison helper)
